### DISCUSSION TITLE
after updating react 18, i have problems with date-fns

### DISCUSSION THREAD

**Discussion by Musho-T** [vercel/next.js#39425] [Help]
### Verify canary release

- [X] I verified that the issue exists in Next.js canary release

### Provide environment information

 "react": "18.1.0",
 "next": "12.1.6",
 "react-dom": "18.1.0",
"date-fns": "2.28.0",

### Describe the Bug

I get errors after the update, this is related to the time zone, there are countries where they do not receive these errors)

`
react-dom.production.min.js:119 Uncaught Error: Minified React error #425; visit https://reactjs.org/docs/error-decoder.html?invariant=425 for the full message or use the non-minified dev environment for full errors and additional helpful warnings.
    at d7 (react-dom.production.min.js:119:38)
    at gt (react-dom.production.min.js:206:80)
    at h6 (react-dom.production.min.js:281:80)
    at h5 (react-dom.production.min.js:280:438)
    at h3 (react-dom.production.min.js:280:315)
    at h2 (react-dom.production.min.js:280:172)
    at hT (react-dom.production.min.js:268:193)
    at A (scheduler.production.min.js:14:181)
    at MessagePort.g (scheduler.production.min.js:15:101)
`

`
react-dom.production.min.js:148 Uncaught Error: Minified React error #418; visit https://reactjs.org/docs/error-decoder.html?invariant=418 for the full message or use the non-minified dev environment for full errors and additional helpful warnings.
    at fp (react-dom.production.min.js:148:277)
    at _ (react-dom.production.min.js:293:444)
    at h5 (react-dom.production.min.js:280:375)
    at h3 (react-dom.production.min.js:280:315)
    at h2 (react-dom.production.min.js:280:172)
    at hT (react-dom.production.min.js:268:193)
    at A (scheduler.production.min.js:14:181)
    at MessagePort.g (scheduler.production.min.js:15:101)
`

`
react-dom.production.min.js:293 Uncaught Error: Minified React error #423; visit https://reactjs.org/docs/error-decoder.html?invariant=423 for the full message or use the non-minified dev environment for full errors and additional helpful warnings.
    at _ (react-dom.production.min.js:293:137)
    at h5 (react-dom.production.min.js:280:375)
    at h3 (react-dom.production.min.js:280:315)
    at h2 (react-dom.production.min.js:280:172)
    at hU (react-dom.production.min.js:271:69)
    at hT (react-dom.production.min.js:268:409)
    at A (scheduler.production.min.js:14:181)
    at MessagePort.g (scheduler.production.min.js:15:101)
`

### Expected Behavior

...

### To Reproduce

on react version 17.0.2 didn't get such errors

**Comment by hamlim** [comment]
Could you share a snippet of code that replicates the issues you are seeing?

Based off of the errors and the context I'd assume that the issue is you're formatting a date on the server and then when the client goes to hydrate into that rendered content the format is different (due to timezones or difference in clocks) leading to a mismatch.

There are a few ways you could resolve this:

- Wait to format the date on the client instead of on the server
- `suppressHydrationWarning` on the nearest parent HTML element around the formatted time _may_ work, although I haven't tested this
- You may also be able to wrap the formatted time in a `<Suspense>` boundary, which React may be able to patch up without errors on the client (also haven't verified this approach)

**Comment by rolandtoth** [comment]
Exactly as @hamlim wrote, we just had such issue with dates. We were outputting only the year of a date and because of server vs client timezone difference it was 1990 vs 1989. Compare the page source and what's actually rendered where it happens - for us it did not happen on localhost (different data).

This was not a new issue but only after updating to React 18 brought it to surface.

**Comment by Musho-T** [comment]
@rolandtoth I will get it only on the server. there are no such errors on localhost

**Comment by khengsaurus** [comment]
Hello @Musho-T can I check how you managed to identify `date-fns` as the cause of your issue?
I'm seeing the exact same errors (425, 418, 423) on vercel deployment, and I'm not using `date-fns`. No such errors when running `yarn dev` & `yarn build && yarn start` locally.

Dependencies:
```json
"next": "^12.1.6",
"react": "^18.2.0",
"react-dom": "^18.2.0",
```

Error logs for #425. Similar error logs for #418 and #423
```
Uncaught Error: Minified React error #425; visit https://reactjs.org/docs/error-decoder.html?invariant=425 for the full message or use the non-minified dev environment for full errors and additional helpful warnings.
    at Zr (framework-4556c45dd113b893.js:1:43705)
    at qu (framework-4556c45dd113b893.js:1:85317)
    at ks (framework-4556c45dd113b893.js:1:109548)
    at bs (framework-4556c45dd113b893.js:1:109429)
    at gs (framework-4556c45dd113b893.js:1:109296)
    at vs (framework-4556c45dd113b893.js:1:109159)
    at ls (framework-4556c45dd113b893.js:1:104263)
    at S (framework-4556c45dd113b893.js:1:138450)
    at MessagePort.T (framework-4556c45dd113b893.js:1:138984)
Zr @ framework-4556c45dd113b893.js:1
qu @ framework-4556c45dd113b893.js:1
ks @ framework-4556c45dd113b893.js:1
bs @ framework-4556c45dd113b893.js:1
gs @ framework-4556c45dd113b893.js:1
vs @ framework-4556c45dd113b893.js:1
ls @ framework-4556c45dd113b893.js:1
S @ framework-4556c45dd113b893.js:1
T @ framework-4556c45dd113b893.js:1
```

**Comment by Moranilt** [comment]
I have the same issue after upgrading React version to 18.2.0 with following errors:
```
framework-4556c45dd113b893.js:1 Uncaught Error: Minified React error #418; visit https://reactjs.org/docs/error-decoder.html?invariant=418 for the full message or use the non-minified dev environment for full errors and additional helpful warnings.
    at sa (framework-4556c45dd113b893.js:1:48561)
    at xi (framework-4556c45dd113b893.js:1:122126)
    at bs (framework-4556c45dd113b893.js:1:109368)
    at gs (framework-4556c45dd113b893.js:1:109296)
    at vs (framework-4556c45dd113b893.js:1:109159)
    at ls (framework-4556c45dd113b893.js:1:104263)
    at S (framework-4556c45dd113b893.js:1:138450)
    at MessagePort.T (framework-4556c45dd113b893.js:1:138984)
 ```
    
    ```
    Uncaught Error: Minified React error #423; visit https://reactjs.org/docs/error-decoder.html?invariant=423 for the full message or use the non-minified dev environment for full errors and additional helpful warnings.
    at xi (framework-4556c45dd113b893.js:1:121812)
    at bs (framework-4556c45dd113b893.js:1:109368)
    at gs (framework-4556c45dd113b893.js:1:109296)
    at vs (framework-4556c45dd113b893.js:1:109159)
    at as (framework-4556c45dd113b893.js:1:105939)
    at ls (framework-4556c45dd113b893.js:1:104490)
    at S (framework-4556c45dd113b893.js:1:138450)
    at MessagePort.T (framework-4556c45dd113b893.js:1:138984)
    ```
    
This error can be reproduced after deploying to a DockerImage with Node v16.

**Comment by balazsorban44** [comment]
Anyone facing this issue, please provide a reproduction repository so we can investigate, thanks!

**Reply by BrooceLR** [reply]
Please consider the actual **underlying issue**: these errors are incomprehensible once minified, but they don't show up at all in development. So it's *very* hard to get a repro.
Please consider[ this discussion ](https://github.com/vercel/next.js/discussions/36641)about React's `onRecoverableError`, which would make significantly easier to get a grip on what is causing the errors and provide repros.

**Comment by malekpouri** [comment]
Hi, I also use the versions:
Dependencies:

```
"next": "^12.1.6",
"react": "^18.2.0",
"react-dom": "^18.2.0",
```
And after start the project, I encountered the following errors:

Uncaught Error: Minified React error #425;
Uncaught Error: Minified React error #418;
Uncaught Error: Minified React error #423;

**Comment by balazsorban44** [comment]
Please read https://github.com/vercel/next.js/issues/37489#issuecomment-1166494166

**Comment by guilleigmu** [comment]
Problem seems to be related with Dates (not `date-fns` only). I've also been working on a project using react-18 and next.js 12.2.0 and we've only been facing this issue since we added a native `Date` object in our pages. 

Looking forward to have some spare time to prepare a little repo example for you @balazsorban44, but I believe any nextjs project using react-18 and has any kind of Date in some of its static pages will do the trick.

**Comment by Musho-T** [comment]
we get this error because the date is different on the server and different on the client)

**Comment by olaj** [comment]
I also have this issue. I managed to fix some of them that was clearly date related but i still get the error on almost all pages but only for mobile users.

The thing that makes it very hard, and time consuming to debug is that no matter what I've tried i can only reproduce it in production. I've tried to run the docker locally and so on but i can't reproduce it :/

You can see the issue if you go to https://alternativeto.net, open the inspector and use the "mobile emulator", then if you refresh the page it throws these errors.

Is it possible to get any hint from the console there? It seems to only be react code and no hints for poor me.

**Comment by churchinsydney-dev** [comment]
I have the same problem. Very strange. I can't reproduce it on my local. 

Problem site: https://churchinsydney-3mtz6jju3-churchinsydney.vercel.app/

The error is gone after removing date-fns 

https://churchinsydney-tefmcruxs-churchinsydney.vercel.app/

```
Uncaught Error: Minified React error #425; 
visit https://reactjs.org/docs/error-decoder.html?invariant=425
 for the full message or use the non-minified 
dev environment for full errors and additional helpful warnings.
```

**Comment by hafffe** [comment]
@churchinsydney-dev your problem is probably difference on server and client. The Server probably uses UTC but not your client, I can spot the difference on "General Announcement" post where on client it says "Updated 2 hours ago" but if I view page source it says "Updated 2 minutes ago" and therefore react complains about different html on server and client.

I would recommend use the package `date-fns-tz` to format your dates if you want to keep using date-fns.

**Comment by iMarv** [comment]
@hafffe die you actually fix the issue with `date-fns-tz`?
I have been hardcoding the timezone everywhere I could find and the page still seems to readjust the dates in the client:

```ts
export function formatDateFull(date: string): string {
  const d = utcToZonedTime(date, "Europe/Berlin");

  return format(d, "iiiiii. do MMMM H:mm", {
    locale: de,
    timeZone: "Europe/Berlin"
  });
}
```

**Comment by zhex900** [comment]
I have the same problem. This is hard to debug since it only happens in the production build.

**Comment by Tonyce** [comment]
also have the same problem(Uncaught Error: Minified React error #425)  without use date-fns. https://dblog-three.vercel.app/

**Comment by zhex900** [comment]
I am not sure the problem is timezone. 

For my site. The static generated site has the value `Last updated:  6 days ago`, but in the live site the value is `Last updated:  3 days ago`

**Comment by Christian-Mar** [comment]
I had the same problem (Minified React errors #425, #418 and #423) with next.js getting the date from Sanity (backend) when I used the following code: 

`<p>Published on {new Date(post._createdAt).toLocalDateString()}</p>`

I installed moment (library): 

`npm install moment`

Imported it: 

` import Moment from 'moment';`

and now use the following code: 

`<p>Published on {Moment(post._createdAt).format('DD-MM-YYY')}<p>`

... and now it works without the errors. 

So, instead of creating a date object with the new Date() constructor, I now just read the date from the backend. 

Maybe this can help for some of you?

**Comment by axclever** [comment]
moment.js solved issue for particular situations but this error still present on pages where I need to render Date multiple times. Like a list of comments with creation date.

Keep diggin

**Comment by axclever** [comment]
You can also doesn't render Date on server and only on client. Maybe it would be helpful for you guys https://github.com/vercel/next.js/discussions/35773?sort=top#discussioncomment-2485078

**Comment by balazsorban44** [comment]
So based on the responses, the issue seems to be that React now warns about hydration mismatches when a `Date` is rendered differently on the server and the client. This is not new to React 18, although the warning has been surpassed before.

This can happen for example if your server is not in the same timezone as your user, rendering the values differently. (That's why you don't experience this locally, since your machine is both the server and the client, at the same time zone.)

An example of this could be:

```jsx
export default function Home(props) {
  return <p>{new Date(props.date).toString()}</p>
}

export function getStaticProps() {
  return { props: { date: Date.now() } }
}
```

There are different techniques to avoid this problem, most of them involve `useEffect` that is only executed after hydration. You can render the server value first, then re-render on the client, you can render a placeholder UI first, or can server-render the page.

Since this is not a Next.js bug, I'll move this to a discussion.

**Comment by JamesSingleton** [comment]
My issue with this is I have a component like this where `dateString` could look like `2021-12-28T20:00:11.170Z` coming from my CMS but because I do `parseISO` which I believe relies on the client's local time zone. But not sure how to resolve this as I use `getStaticProps` for almost everything

```tsx
import { parseISO, format } from 'date-fns'

import type { WithClassName } from '@types'

interface DateProps extends WithClassName {
  dateString: string
}

export default function Date({ dateString, className }: DateProps) {
  const date = parseISO(dateString)

  return (
    <time dateTime={dateString} className={className}>
      {format(date, 'LLL d, yyyy')}
    </time>
  )
}
```

**Reply by eliseumds** [reply]
@JamesSingleton have you tried `<time dateTime={dateString} className={className} suppressHydrationWarning>`?

**Reply by zhex900** [reply]
My workaround is to prevent static generation on the server side. 

```js
  import { formatDistanceToNow } from 'date-fns';

  const [lastUpdated, setLastUpdated] = useState('');

  useEffect(() => {
    // prevent server side rendering
    if (post.dateUpdated && window) {
      setLastUpdated(formatDistanceToNow(new Date(post.dateUpdated)));
    }
  }, [post.dateUpdated]);
```

The solution is to force date logic to the client side. This makes sense since the date of static generation is going to be different to the client. I assume this applies to server side rendering too.

**Reply by MarioGiancini** [reply]
@eliseumds @JamesSingleton @zhex900 @xandor-io 

I was also able to fix the Hydration errors by using `suppressHydrationWarning` on `<time>`.

Was getting these in the console:
```
Uncaught Error: Minified React error #425;
Uncaught Error: Minified React error #418;
Uncaught Error: Minified React error #423;
```

I noted the number of console errors was related to the number of `<time>` tags in the source. I had this as my Date component:

```typescript
import { format, parseISO } from 'date-fns'

export default function Date({ dateString }: { dateString: string }) {
  if (!dateString) return null

  const date = parseISO(dateString)
  return <time dateTime={dateString}>{format(date, 'LLLL d, yyyy')}</time>
}
```

I changed it to this and errors were gone on the next deployment:
```typescript
import { format, parseISO } from 'date-fns'

export default function Date({ dateString }: { dateString: string }) {
  if (!dateString) return null

  const date = parseISO(dateString)
  // @see https://github.com/vercel/next.js/discussions/39425#discussioncomment-3367041 
  return <time dateTime={dateString} suppressHydrationWarning>{format(date, 'LLLL d, yyyy')}</time>
}
```

Environment (node **v16** on vercel):
```
"node": "^16.18.1"
"next": "^13.0.2",
"react": "^18.2.0",
"react-dom": "^18.2.0",
```

**Reply by JamesSingleton** [reply]
Yeah that worked for me as well... Just annoying to have to do it

**Reply by MarioGiancini** [reply]
Agreed.

**Comment by imhuynq** [comment]
This issue also happened with dayjs timezone

**Comment by Haessler16** [comment]
So in my case i fix this issue putting false to the swcMinify variable in the next.config file.
`const nextConfig = {
   swcMinify: false
}`

**Reply by adikancherla** [reply]
this worked for me too

**Comment by ghost** [comment]
With the input of @balazsorban44 and @zhex900, I was able to avoid this error for the time being.

```jsx
import { format } from 'date-fns'

export default function Post({ created_at }) {
  const [createdAt, setCreatedAt] = useState(created_at)

  useEffect(() => {
    if (created_at) {
      setCreatedAt(format(new Date(created_at), 'h:mm a ⸱ MMM d, y'))
    }
  }, [created_at])

  return <time dateTime={created_at}>{createdAt}</time>
}
```

**Reply by l-white** [reply]
Normally a great solution, but we took a hit on performance with a similar implementation

**Comment by MikeCheek** [comment]
I found another code mistake that causes minified error #418 in react

I had minified react error #418 in my GatsbyJs website

I finally found the cause on my site and it's not connected with a date issue

The problem was that I had some components lazy loaded but I forgot to wrap them in a Suspense tag

```javascript
// will cause the error
const Component= lazy(() => import('../../atoms/component'))
{...}
return <Component/>


// ok
const Component= lazy(() => import('../../atoms/component'))
{...}
return <Suspense fallback={null}>
    <Component/>
 </Suspense>
```

Even in this case, the problem is that the rendered version of the site is different from the server one

I hope this will help someone

**Comment by xandor-io** [comment]
Im having this issue currently when I deploy to vercel. (I'm fairly new with vercel and next.js) 

Uncaught Error: Minified React error https://github.com/vercel/next.js/pull/425;
Uncaught Error: Minified React error https://github.com/vercel/next.js/issues/418;
Uncaught Error: Minified React error https://github.com/vercel/next.js/pull/423;

Anyone know best way to debug errors in production build? Is there a flag I can set in `next.config.js` to show me react errors in production? Any help will be much appreciate it since I cant figure this out. 

```
"next": "12.3.1",
"react": "18.2.0",
"react-dom": "18.2.0",
```

**Reply by songtuyen97** [reply]
Because the issue relate to display time.
You can reproduce the issue on local by
1. set another timezone in source.
set `process.env.TZ = "America/Denver";` to root component in `_app.tsx` 

2. Run nextjs in docker with another timezone, then access normal.

**Comment by songtuyen97** [comment]
Because the issue relate to display time.
You can reproduce the issue on local by ways:
1. set another timezone in source.
set `process.env.TZ = "America/Denver";` to root component in `_app.tsx` 
`yarn build & yarn start`

2. Run nextjs in docker with another timezone, then access normal.

**Comment by gtakat** [comment]
I have the same problem. I am using `dayjs`.

In my environment, I have Japan time (`YYYY-MM-DDTHH:mm:ss+09:00`) in the variable and set `utc` on the server side.

I was able to resolve the error by setting the timezone.

#### before (has error)

```javascript
dayjs(date).format(...)
```

#### after (no error)

```javascript
dayjs.extend(utc)
dayjs.extend(timezone)
dayjs.tz.setDefault('Asia/Tokyo')
dayjs.locale('ja')

dayjs(date).tz().format(...)
```

**Reply by AbelPark** [reply]
cool

**Reply by Ibadichan** [reply]
instead of  `dayjs(date).tz().format(...)` use `dayjs.tz(date).format(...)`

**Comment by dejurin** [comment]
This problem without date-fns
<img width="1369" alt="Screenshot 2023-02-03 at 20 45 26" src="https://user-images.githubusercontent.com/4470015/216682801-9e7fafca-83da-4dbd-9d5a-9065fe0fc533.png">

**Comment by rubby-c** [comment]
I have the same problem, I'm storing an UTC timestamp long in my database, fetching it from getServerSideProps and using `new Date(timestamp).toLocaleString()`.

I have several dynamic pages and the errors only show on a few of them even though everything related to time is the same. The issue only occurs on my Linux server.

```js

// time.js

let options = { year: 'numeric', month: 'long', day: 'numeric' };

export function ToIso(timestamp) {
    return new Date(timestamp).toISOString();
}

export function ToDate(timestamp) {
    return new Date(timestamp).toLocaleDateString('en-US', options);
}

// [post].js

<time itemProp='datePublished' dateTime={ToIso(post.timestamp)}>{ToDate(post.timestamp)}</time>
```

**Reply by rubby-c** [reply]
Figured it out! In my case it was Cloudflare HTML minifying that produced these errors.

**Reply by beniven** [reply]
@rubby-c Thank you! Disabling Auto Minify for HTML in Cloudflare eliminated the hydration errors in our case as well.

**Reply by vamcs** [reply]
Thank you so much!!! I've spent months debugging this in different ways and this was exactly it. Our Swedish site had the issue but not the Norwegian, Finnish and Danish, because only the SE version had the minify activated.

For anyone interested:
Log in to Cloudflare > Choose your domain > "Speed" > "Optimization" > "Content Optimization" > Uncheck everything under "Auto Minify"

**Comment by brunocascio** [comment]
I faced this issue too (nextjs and remix)

In my case was the chrome extensions that mutate the dom (e.g. Dashlante, LastPass, etc)

**Reply by liam-russell** [reply]
FYI I logged the LastPass issue here: https://github.com/vercel/next.js/issues/51816
Please upvote it if you have come across this issue yourself

**Comment by peasandwell** [comment]
We were experiencing the same issue, as mentioned in another comment on this discussion the problem was that we were rendering dates using the `'en-GB'` format, whereas the server was using the default timezone. This meant that the server was rendering "UTC", whereas the client was trying to render "GMT".

I only spotted this when I added `suppressHydrationWarning`, I could spot a quick flash where "UTC" changed to "GMT" when I was refreshing the page.

We were using docker to handle our deployments, so our fix was to add `ENV TZ="Europe/London"` to our `dockerfile` (so underneath line 36 on https://github.com/vercel/next.js/blob/canary/examples/with-docker/Dockerfile):

```
# Production image, copy all the files and run next
FROM base AS runner

# Override default timezone, preventing hydration errors when rendering timezones
ENV TZ="Europe/London"

WORKDIR /app
```

This works for us as our users will only be UK based, however we're yet to see if this has any implications though.

**Reply by kelvinkaicheung** [reply]
In my case, I used a toUTCString in getServerSideProps to pass it to the client and then convert back to date object. After changing toUTCString() to toISOString() the error messages disappear. Proven that you are correct!

**Comment by madfcat** [comment]
It's definitely a React related issue. For me, it happens in a Gatsby project only after I deploy to S3 (aws). I have these versions:
```
    "gatsby": "^5.2.0",
    "react": "^18.2.0"
```

Happens without Date being used in the whole project...

Update: I think it was something related to the translation plugin (gatsby-plugin-react-i18next)

**Comment by gabrielmaldi** [comment]
Expanding on @songtuyen97's comment, I was able to diagnose my issue by running:

```sh
TZ="America/Denver" npm run dev # TZ must be different to your actual timezone
```

This produces a nice, focused and reproducible error in my local development environment:

<img width="873" alt="Screenshot 2023-07-03 at 11 18 03" src="https://github.com/vercel/next.js/assets/3728897/3169ecfb-21a5-41c5-a556-5ab3e7dfdab0">

**Comment by rhirayamaaan** [comment]
I encountered the same problem as well, but it has been resolved. I made use of @gtakat ’s comment. https://github.com/vercel/next.js/discussions/39425#discussioncomment-4409455
I’m using the date-fns library, so I installed data-fns-tz and used `utcToZonedTime` and `format`.

before:
```tsx
import { FC, useMemo } from 'react';
import utcToZonedTime from 'date-fns-tz/utcToZonedTime';
import format from 'date-fns-tz/format';

type Props = {
  zone?: 'Asia/Tokyo';
  children?: Date;
  pattern?: 'yyyy.M.dd';
};

const DateFormat: FC<Props> = ({
  zone = 'Asia/Tokyo',
  children,
  pattern = 'yyyy.M.dd',
}) => {
  const formattedDate = useMemo(() => {
    if (!children) {
      return undefined;
    }

    const date = utcToZonedTime(children, zone);

    return format(date, pattern);
  }, [children, zone, pattern]);
  return <>{formattedDate}</>;
};
```

after:
```tsx
...
  const formattedDate = useMemo(() => {
    if (!children) {
      return undefined;
    }

    const date = utcToZonedTime(children, zone);

    return format(date, pattern, {
      timeZone: zone,
    });

  }, [children, zone, pattern]);
...
```

I forgot to specify the third argument of `format`.

@gtakat Thank you for your comment! ありがとうございます！大変助かりました！

**Reply by rhirayamaaan** [reply]
I’m sorry… This solution has a problem.
I'll try to resolve this problem.
I'll comment again when I know more...

**Comment by smebberson** [comment]
If you want to render the date in the timezone of the user keep everything on the server in UTC and render the date on the client only. Try this:

- Run your server in UTC (and your local Node.js development in UTC (`TZ=UTC node ./index.js` or `$env:TZ = 'UTC'` on windows)).
- Keep your dates in UTC in the database and everywhere else.

Create a client component that renders dates on the client:

```javascript
'use client';

export const dateFormat = ({ date }) =>
    date.toLocaleString(
        // eslint-disable-next-line no-undefined
        undefined,
        {
            year: 'numeric',
            month: 'short',
            day: 'numeric',
        }
    );

export default function RenderDate({ date }) {
    return <span>{dateFormat({ date: new Date(date) })}</span>;
}
```

Create a server component that retrieves dates and uses next/dynamic to import the client side component ensuring that SSR doesn't take place:
```javascript
import dynamic from 'next/dynamic';

const RenderDate = dynamic(() => import('./date'), {
    ssr: false,
});

const getData = async () => {

    // replace this with something that retrieves dates
    return new Date();

}

export default async function Page() {

    const date = await getData();
    
    return <div><RenderDate date={date} /></div>;

}
```

It should all work and keeps working with dates easy to reason about and simple.

**Comment by 7sferry** [comment]
in my case this happened because of timezone problem. The times changed quickly from UTC to timezone I used everytime I refresh the page. I resolved this by re-parse date and defining timezone I used at date variables. for example I used Jakarta TimeZone.

```ts
const reparseDate = (date: Date) => date.toLocaleString("en-US", { timeZone: "Asia/Jakarta" });

const getPublishDate = (date: Date) => format(reparseDate(date), "MMMM do, yyyy");

```
Code above works in my case

alternative trick I used when using distance time in date-fns: don't show text from server side

in component function:
```ts
const [isClient, setIsClient] = useState(false);

  useEffect(() => {
    setIsClient(true);
  }, []);
```
and in rendering section:
```ts
<span >{`updated ${isClient ? formatDistanceToNow(updatedAt) : "…"} ago`}</span>
```

**Reply by nelsonfrank** [reply]
Thank you @7sferry, this work for me easily after struggle for long time. I use react18, date-fns 3.6.0
